### PR TITLE
Several system roles & small fixes

### PR DESCRIPTION
**Pull request by rimnvd**

### CODE REVIEW

**Comment by IP-13** [comment]
|Overall Project|61.28% **`-1.93%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|18.86%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[MiscSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FMiscSecurityExpression.java)|98.52%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|68.25%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|61.83% **`-13.58%`**|:x:|
|[AuthenticationService.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FAuthenticationService.java)|15.79% **`-0.75%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.11% **`-3.86%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/39300d1ff5c692d8848dc32fc83ac0d8dbf4161e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|6.67% **`-81.11%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|61.28% **`-1.93%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|18.86%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[MiscSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FMiscSecurityExpression.java)|98.52%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|68.25%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|61.83% **`-13.58%`**|:x:|
|[AuthenticationService.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FAuthenticationService.java)|15.79% **`-0.75%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.11% **`-3.86%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/a0349cdfbe19d6a95f7db47c65c57734aedc0edb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|6.67% **`-81.11%`**|:x:|

File: src/main/java/org/itmo/eventapp/main/security/securityexpression/RoleSecurityExpression.java [hunk 1]
```diff
@@ -2,102 +2,62 @@
...
 
     public boolean canAssignOrganizerRole(@Positive(message = "Параметр eventId не может быть меньше 1!") int eventId) {
-        return checkEventPrivilege(eventId, PrivilegeName.ASSIGN_ORGANIZER_ROLE) || checkSystemPrivilege(PrivilegeName.ASSIGN_ORGANIZER_ROLE);
+        return miscSecurityExpression.checkEventPrivilege(eventId, PrivilegeName.ASSIGN_ORGANIZER_ROLE);
     }
 
     public boolean canGetAllOrganizationalRole(@Positive(message = "Параметр eventId не может быть меньше 1!") int eventId) {
         return canAssignOrganizationalRole(eventId) || canAssignOrganizerRole(eventId) ||
-            canAssignAssistantRole(eventId) || canRevokeAssistantRole(eventId) || canRevokeOrganizationalRole(eventId);
+                canAssignAssistantRole(eventId) || canRevokeAssistantRole(eventId) || canRevokeOrganizationalRole(eventId);
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/security/securityexpression/RoleSecurityExpression.java:61]: А тут не должно быть еще одного canRevokeOrganizerRole?

File: src/main/java/org/itmo/eventapp/main/service/RoleService.java [hunk 2]
```diff
@@ -108,16 +108,27 @@ public void assignSystemRole(Integer currentPrincipalId, Integer userId, Integer
...
             throw new ResponseStatusException(HttpStatus.FORBIDDEN, ExceptionConst.REVOKE_SELF_ROLE_FORBIDDEN_MESSAGE);
         var user = userService.findById(userId);
-        user.setRole(getReaderRole());
-        userService.save(user);
+        var role = findRoleById(roleId);
+        if (role.getType().equals(RoleType.EVENT))
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, ExceptionConst.INVALID_ROLE_TYPE.formatted("системная"));
+        var roles = user.getRoles();
+        if (roles.size() == 1) {
+            throw new ResponseStatusException(HttpStatus.FORBIDDEN, ExceptionConst.AT_LEAST_ONE_SYSTEM_ROLE_MESSAGE);
```
[the remark this answers is not in the document]
Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Ну можно попытаться и роль читателя удалить, я бы не стала, как будто бы это не очевидно

File: src/test/resources/db/test-migration/V0_2__create_tables_test.sql [hunk 3]
```diff
@@ -37,11 +37,16 @@ create table if not exists user_notification_info(
 create table if not exists user_t
 (
     id int generated always as identity primary key not null,
-    role_id int not null references role(id),
     notification_info_id int not null references user_notification_info(id),
     name varchar(256) not null,
     surname varchar(256)
 );
+create table if not exists user_role
```
Comment (IP-13) [inline src/test/resources/db/test-migration/V0_2__create_tables_test.sql:44]: еще бы в тест добавить очещение этой таблицы. В clean_tables.sql

File: src/main/java/org/itmo/eventapp/main/security/securityexpression/RoleSecurityExpression.java (hunk 1 above)
Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/security/securityexpression/RoleSecurityExpression.java:61]: Это привилегия админа, не привязанная к мероприятию. Возможно, это придется переделать из-за того, что мы переносим назначение орг ролей в другое место, но пока так

**Comment by IP-13** [comment]
|Overall Project|61.47% **`-1.93%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|18.86%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[MiscSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FMiscSecurityExpression.java)|98.52%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|68.25%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|61.83% **`-13.58%`**|:x:|
|[AuthenticationService.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FAuthenticationService.java)|15.79% **`-0.75%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.11% **`-3.86%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/c6e4f6f4c7ec046f6990d5c6db8fe063dc4842f6/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|6.67% **`-81.11%`**|:x:|

File: src/test/resources/db/test-migration/V0_2__create_tables_test.sql (hunk 3 above)
Comment (rimnvd) [reply src/test/resources/db/test-migration/V0_2__create_tables_test.sql:44]: А это не гарантирует удаление данных из зависимых таблиц? 
**truncate table user_t cascade;**

Comment (rimnvd) [reply src/test/resources/db/test-migration/V0_2__create_tables_test.sql:44]: Если верить документации Postgres, то **truncate table user_t cascade;** должна очищать и таблицы, ссылающиеся на нее по внешнему ключу

Comment (IP-13) [reply src/test/resources/db/test-migration/V0_2__create_tables_test.sql:44]: Ну для консистентности лучше дописать, а то это не очевидно

File: src/main/java/org/itmo/eventapp/main/service/RoleService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Получается, что нельзя забрать роль, если у человека только одна роль. Тогда может быть админ, который удалит себе все роли и начнет делать плохие вещи, и никто не сможет ему помешать

[the remark this answers is not in the document]
Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Хотя другой админ может назначить ему роль читателя, а потом удалить роль админа

[the remark this answers is not in the document]
Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Но это уже игры с бубном какие-то

[the remark this answers is not in the document]
Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Удалять у себя системные роли нельзя, организационные можно

[the remark this answers is not in the document]
Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:126]: Назначать тоже

**Comment by IP-13** [comment]
|Overall Project|61.47% **`-1.93%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|18.86%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[MiscSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FMiscSecurityExpression.java)|98.52%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|68.25%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|61.83% **`-13.58%`**|:x:|
|[AuthenticationService.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FAuthenticationService.java)|15.79% **`-0.75%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.11% **`-3.86%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/461ccf748e5219e83a0d1e1dcb02b87eb9904d87/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|6.67% **`-81.11%`**|:x:|

File: src/test/resources/sql/clean_tables.sql [hunk 4]
```diff
@@ -26,7 +26,8 @@ alter sequence user_login_info_id_seq restart with 1;
 alter sequence user_t_id_seq restart with 1;
 alter sequence user_notification_info_id_seq restart with 1;
 alter sequence registration_request_id_seq restart with 1;
-alter sequence role_privilege_id_seq restart with 63;   -- Number of roles-privilege relations = 62
-                                                        -- Admin (22) ; Reader (6) ; Organizer (25) ; Assistant (9) ;
+alter sequence user_role_id_seq restart with 1;
+alter sequence role_privilege_id_seq restart with 63;   -- Number of roles-privilege relations = 61
```
Comment (IP-13) [inline src/test/resources/sql/clean_tables.sql:30]: тут 62 надо сделать?

**Review by IP-13 — APPROVED** [review]

**Pull request merged**